src/org/freedesktop/gstreamer/Pad.java

```diff
@@ -318,9 +328,45 @@ public void block(final Runnable callback) {
             public PadProbeReturn eventReceived(Pad pad, Event event) {
                 callback.run();
                 pad.removeCallback(EVENT_PROBE.class, this);
-                return PadProbeReturn.DROP;
+                return PadProbeReturn.REMOVE;
             }
-        }, GstPadAPI.GST_PAD_PROBE_TYPE_BLOCKING | GstPadAPI.GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM);
+        }, GstPadAPI.GST_PAD_PROBE_TYPE_IDLE);
```

Comment:
I'm concerned this is changing the semantics of an existing method. Even if it currently has some issues. I'm not against it, as it's probably the right change. Are there any uses that might break with this change?

I am inclined to fix `block()` but also deprecate it with a view to adding generic `addProbe()` support. The idea of gst1-java-core is to stick close to implementing upstream GStreamer behaviour rather than add to it.

Having said that, wonder whether to deprecate `block()` and add a new method with your fixes - `onIdle(Runnable r)` ?

Reply:
According to the Javadoc, `.block` does "Run a runnable under a blocked state".
GST_PAD_PROBE_TYPE_IDLE does that as well as GST_PAD_PROBE_TYPE_BLOCKING with the only difference, that GST_PAD_PROBE_TYPE_BLOCKING will only be called as soon as there is actual data flowing on a Pad (and the Data-Pushing Thread is then calling the Probe) whereas GST_PAD_PROBE_TYPE_IDLE can also be called directly from the probe-add.

The implication of this is, that a Probe for GST_PAD_PROBE_TYPE_BLOCKING might *never* be called, when there is no actual dataflow (ie. the source is some kind of network source, lets say an udpsrc) and you want to unlink it, then blocking for GST_PAD_PROBE_TYPE_BLOCKING might not work if the sending side does not send anymore packets and your probe will never be called.

Before GST_PAD_PROBE_TYPE_IDLE was introduced, one had to manually inject an event into the pipeline force a callback for GST_PAD_PROBE_TYPE_BLOCKING even when no data is flowing.

To to answer your question: I think that only blocking for GST_PAD_PROBE_TYPE_IDLE does exactly what the Javadoc and Method-Name indicates and what the caller expects.

Reply:
I think you're mixing up `GST_PAD_PROBE_TYPE_BLOCKING` and `GST_PAD_PROBE_TYPE_BLOCK`. Now, I'm not sure if anyone could be relying on the behaviour of the former.

Reply:
You are correct, I mixed them up in my comment.

In the usage `GstPadAPI.GST_PAD_PROBE_TYPE_BLOCKING | GstPadAPI.GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM` the Mask did match both, BLOCK and IDLE, which would then possibly call the Runnable twice.

Reply:
Yes, my thought too, and why is the event mask there?!  This was added before I took on the fork, but it does seem wrong - let's go with your change here and if anyone complains we can rethink it!